### PR TITLE
UI Add Cluster Activity Page

### LINKED ISSUE #16081: Add dashboard to see DAGs (system wide status) overview
Airflow is great in a specific DAG checks. Meaning if you know there is an issue with a specific DAG you have everything you need to check what is going on but if you have a system of 1000+ dags your needs are a bit different.
It's missing some kind of dashboard that will show what is going on in system wise manner.
For example:
TOP X dags that are still in running mode
Pools that are max occupied
last scheduler heartbeat
I can think of many other metrics

The idea here is a dashboard that will show what is going on in Airflow right now (not calculating over history).

There are many ways to get it externally with Grafana and other tools but I think it comes to a time where Airflow should provide it itself.

### PR DESCRIPTION
**Pull request by pierrejeambrun**
closes: #16081

This add a native dashboard page to the UI to gather a few useful metrics to monitor your airflow cluster.

Up to now people had to build their own dashboard with third parties tools such as grafana.

This PR is not finalized yet but aim to gather feedback on the general structure, UI or any other improvement you might want to see in this page before I go ahead and move forward with a more complete implementation.

- Add a chart library to our front end, leverage another Apache project [Echart](https://github.com/apache/echarts) :tada:  (which is great btw, very flexible and exhaustive library).
- 2 kind of metrics (**live** that uses current data, and **historical** that fetch older records from the db and has filter enabled)
- Everything is auto refreshed
- Live metrics use different REST API calls to fetch information that we have available
- Historical metrics use a single call to a custom private endpoint that fetch everything we need. (Could split that and integrate it to the REST API but we would need some standards for KPIs endpoint and generic way of fetching that to be consistent across resources, i.e how to pass the columns of interest, aggregation type, validation standard payload response etc...). We could do that in a separate PR because this would require on it's own some effort
- UI is responsive and wraps as needed. (Could definitely be improved)

TODO:
- [x] Add tests
- [x] Fix colors by plugin in `STATE_COLORS` and Chakra theme / harmonize the palette
- [x] Factorize some front end components (maybe with compound components)
- [ ] ~~Memoize options if needed~~ (no gain identified)

image
image
image
image

### CODE REVIEW

**Comment by eladkal** [comment]
This looks great! Kudos!

Couple of ideas to consider:
1. Viewing dashboard only for specific owner / specific tags
2. [STATE_COLORS](https://airflow.apache.org/docs/apache-airflow/stable/howto/customize-ui.html#customizing-state-colours) is configurable. can we support the colors user define? (Not sure if this is what you meant in the TODO)
3. As part of the Live metrics would be nice to also see Number of DAGs to be scheduled in the next hour
4. If we can some how show pool utilization across time? e.g graph of occupied slots vs time. This can be useful to detect bottle necks.
5. I think the dashboard can be named: Cluster Activity?

**Comment by potiuk** [comment]
Whoa . This is going to be incredibly useful! 

I think especially if it could focus on Airflow-specific charts and graphs, surfacing some common problems people might have.

Also a small idea sparked from those chart. How about those (way in the future) with some guidance on how to approach seeing some of those values ! 

Think of the instructions for your appliance: If you see 3 blinking dots, it likely means that problem and you can approach it possibly by cleaning your ...  (half-joking of course, but I think there are a number o things that are petty repeatably fixable problems in Airflow if you could surface some problems in forms of similar charts and people would have some instructions next to those charts. 

There is a long way from the charts to such instructions, but I think they are an absolute prerequisite and extremely useful  on its own.

**Comment by bbovenzi** [comment]
Great work! Some quick usability comments:

Let's put some labels here:
Screenshot 2023-05-08 at 12 24 17 PM

Looks like we need to fix some overflow:
Screenshot 2023-05-08 at 12 24 05 PM

Let's use the same state colors as the rest of the UI. I think that's already in your todo though. I also wonder if we can order the legend by number of times that the task instance state is in the pie chart:
Screenshot 2023-05-08 at 12 23 16 PM

File: airflow/www/static/js/dashboard/live-metrics/Pools.tsx [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            <Heading size="md">Pools Slots</Heading>
+          </CardHeader>
+          <CardBody>
+            <Box height="250px">
+              <ReactECharts option={option} />
+            </Box>
+          </CardBody>
+        </Card>
+      ) : (
+        <Spinner color="blue.500" speed="1s" mr="4px" size="xl" />
```
Comment (bbovenzi) [inline airflow/www/static/js/dashboard/live-metrics/Pools.tsx]: If there is an error a user will only get a spinner?

File: airflow/www/static/js/api/useDagRuns.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+const useDagRuns = ({
+  dagId,
+  state,
+  limit,
+  orderBy,
+}: API.GetDagRunsVariables) => {
+  const dagRunsUrl = getMetaValue("dag_runs_url").replace("__DAG_ID__", dagId);
+
+  return useQuery(
+    ["dag", state, dagId, limit],
```
Comment (bbovenzi) [inline airflow/www/static/js/api/useDagRuns.tsx]:
```suggestion
    ["dagRuns", state, dagId, limit],
```

Comment (bbovenzi) [inline airflow/www/static/js/api/useDagRuns.tsx]:
```suggestion
    ["dagRuns", state, dagId, limit],
```

File: airflow/www/static/js/dashboard/nav/FilterBar.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  const { timezone } = useTimezone();
+  const startDate = moment(filters.startDate);
+  const endDate = moment(filters.endDate);
+  const formattedStartDate = startDate.tz(timezone).format(isoFormatWithoutTZ);
+  const formattedEndDate = endDate.tz(timezone).format(isoFormatWithoutTZ);
+
+  const inputStyles = { backgroundColor: "white", size: "lg" };
+
+  return (
+    <Flex backgroundColor="#f0f0f0" mb={4} p={4} justifyContent="space-between">
```
Comment (bbovenzi) [inline airflow/www/static/js/dashboard/nav/FilterBar.tsx]: Do we want to make sure we use colors from theme instead of manual hex codes?

File: airflow/www/static/js/dashboard/useFilters.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+export const now = date.toISOString();
+
+const useFilters = (): FilterHookReturn => {
+  const [searchParams, setSearchParams] = useSearchParams();
+
+  const endDate = searchParams.get(END_DATE_PARAM) || now;
+  const startDate =
+    searchParams.get(START_DATE_PARAM) ||
+    moment(endDate).subtract(1, "d").toISOString();
```
Comment (bbovenzi) [inline airflow/www/static/js/dashboard/useFilters.tsx]: I think it would be helpful to say the range in the filter bar. (Showing data for the past day, week, 6 hours, etc)

**Comment by bbovenzi** [comment]
One more issue, if the nav wraps, the container padding doesn't update and the first item can be cut off.

Screenshot 2023-05-15 at 12 33 52 PM

File: airflow/www/static/js/dashboard/live-metrics/Pools.tsx (hunk 1 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/dashboard/live-metrics/Pools.tsx]: Updated with an error message. Also added a generic component that can display a spinner on fetch, and an error message onError. (Will update the screenshot later to display all these cases)

File: airflow/www/static/js/cluster-activity/historical-metrics/index.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+          <Flex alignItems="center">
+            <Heading size="md">Historical metrics</Heading>
+            <InfoTooltip
+              label="Based on historical data. You can adjust the period by setting a different start and end date filter."
+              size={18}
+            />
+          </Flex>
+        </CardHeader>
+        <CardBody>
+          <FilterBar />
```
Comment (bbovenzi) [inline airflow/www/static/js/cluster-activity/historical-metrics/index.tsx:47]: Can we still put in a human readable duration somewhere here? 
"10 dag runs and 124 task instances in the last {day, week, 4 hours, etc}"

**Comment by pierrejeambrun** [comment]
Thank you for the feedbacks and early reviews, I updated the PR. I implemented everything I could that didn't required too much effort as the change is already considerably big. I would be glad to add more complex features in follow up PRs.

Screenshots have been updated :)

**Comment by eladkal** [comment]
I gathered some feedback from the community (probably ideas for future work)

1. Top 5 dags/tasks that constantly retry (motivation: if task always succesful after 3 retries its hard to notice it and ideally should be invistigated why it never succesed on 1st try)
2. Anomally - show if latest 5 dagrus took segnificantly longer time to run then previous 5 dagruns.
3. Bar graph that shows pool utilization across 24 hours.

**Comment by eladkal** [comment]
Since no room for all statuses in the legened can we prefer to present first statuses that appear in the graph?

For example the graph doesnt show dagruns in Queued  yet Queued appear on the legend. On the other had Succeed appear in graph but not in the legend.

Screenshot_20230520_123706_Chrome

**Comment by pierrejeambrun** [comment]
Thank you for the valuable feedback and ideas.

> Since no room for all statuses in the legened can we prefer to present first statuses that appear in the graph?

Done for all the pie chart. For the Pool ordering in the same way is not straight forward, it a little trickier. (success is missing from one pool, but exists on another pool, etc). But the overflow might never happen for the pool as the graph is really big. I will still try to implement though, I'll let you know.

> I gathered some feedback from the community (probably ideas for future work)
Yes, glad to work on those in follow up PRs.

File: airflow/www/static/js/cluster-activity/live-metrics/Health.tsx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+const StatusRow = ({ status }: { status?: API.HealthStatus }) => (
+  <Stack direction="row">
+    <Text textIndent="15px">status:</Text>
+    <div>
+      <Badge
+        colorScheme={status === "healthy" ? "green" : "red"}
+        fontSize="0.9rem"
+      >
+        {status || "unknown"}
```
Comment (bbovenzi) [inline airflow/www/static/js/cluster-activity/live-metrics/Health.tsx]: If it's unknown, should it still be red?

File: airflow/www/static/js/cluster-activity/nav/FilterBar.tsx [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+      <Flex px={2}>
+        <Button
+          colorScheme="cyan"
+          aria-label="Reset filters"
+          background="white"
+          variant="outline"
+          onClick={clearFilters}
+          size="lg"
+        >
+          Clear Filters
```
Comment (bbovenzi) [inline airflow/www/static/js/cluster-activity/nav/FilterBar.tsx:94]: Should we only show this when the start/end are not the default?

**Review by bbovenzi — APPROVED** [review]
Few minor comments but overall lgtm

I wonder if we could include triggerer status in the future too and avoid this:
Screenshot 2023-05-25 at 3 55 13 PM

**Comment by pierrejeambrun** [comment]
Thanks Brent, I will update soon. Yes I saw that @pankajkoti added the triggered health in the API response https://github.com/apache/airflow/pull/31529. I will add it to the dashboard as well. After that we can remove the banner I think

File: airflow/www/static/js/cluster-activity/live-metrics/Dags.tsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      >
+        <Card w="100%">
+          <CardHeader textAlign="center" p={3}>
+            <Heading size="md">DAGs</Heading>
+          </CardHeader>
+          <CardBody>
+            <Flex flexDirection="column" mb={5}>
+              <Text as="b" color="blue.600">
+                Number of unpaused DAGs:
+              </Text>
```
Comment (jedcunningham) [inline airflow/www/static/js/cluster-activity/live-metrics/Dags.tsx]: I feel like this section might look better without this chunk of text.

File: airflow/www/static/js/cluster-activity/live-metrics/Dags.tsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+              <Flex justifyContent="center" mt={2}>
+                <Heading as="b" size="xl">
+                  {dataOnlyUnpaused?.totalEntries}
+                </Heading>
+              </Flex>
+            </Flex>
+            <Flex justifyContent="end" textAlign="right">
+              <Text size="md" color="gray.500">
+                on a total of <Text as="b">{data?.totalEntries}</Text> DAGs
```
Comment (jedcunningham) [inline airflow/www/static/js/cluster-activity/live-metrics/Dags.tsx]: Maybe reword this section so we don't need it?

```suggestion
                unpaused DAGs out of <Text as="b">{data?.totalEntries}</Text> total DAGs
```

or something.

File: airflow/www/static/js/cluster-activity/live-metrics/DagRuns.tsx [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    <Center {...props}>
+      <LoadingWrapper hasData={!!data} isError={isError}>
+        <Card w="100%">
+          <CardHeader textAlign="center" p={3}>
+            <Heading size="md">Dag Runs</Heading>
+          </CardHeader>
+          <CardBody>
+            <Flex flexDirection="column" mb={5}>
+              <Text as="b" color="blue.600">
+                Top 5 longest Dag Runs to finish:
```
Comment (jedcunningham) [inline airflow/www/static/js/cluster-activity/live-metrics/DagRuns.tsx]: Same here, maybe move the "5 longest" into the heading a few lines up?

File: airflow/www/static/js/cluster-activity/live-metrics/DagRuns.tsx [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+                  >
+                    <Thead>
+                      <Tr>
+                        <Th>Dag Id</Th>
+                        <Th>Run Type</Th>
+                        <Th>Duration</Th>
+                      </Tr>
+                    </Thead>
+                    <Tbody>
+                      {data?.dagRuns?.map((dagRun) => (
```
Comment (jedcunningham) [inline airflow/www/static/js/cluster-activity/live-metrics/DagRuns.tsx]: I'm not sure we should expose dag_ids to viewers if they don't also have read on that specific DAG. Maybe we hide this from viewers instead of trying to filter? Thoughts?

The rest are generic enough I don't think it matters, but this specifically makes me at least want to have the conversation.

File: airflow/www/static/js/cluster-activity/live-metrics/Health.tsx (hunk 6 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/cluster-activity/live-metrics/Health.tsx]: I'm not sure this even has a real use case now that I look at it close. I think `get_airflow_health` cannot return null or underfined status, it's either `healthy` or `unhealthy`. This is just to comply with typescript and handle those cases but might never get displayed. (also I feel like anything that is not 'green' is bad and should be displayed in red, cannot think of a use case where you would want your scheduler or medatabase not in a healthy state).

We might want to update the spec

edit: For the triggerer we can have 'none' state

File: airflow/www/static/js/cluster-activity/nav/FilterBar.tsx (hunk 7 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/cluster-activity/nav/FilterBar.tsx:94]: I tried to keep this consistent with the filter bar in the Grid.
image

Also having elements pop in and out of the page (especially buttons) feels a little weird IMHO. We could grey it out but we would need to do the same in the grid.

File: airflow/www/static/js/cluster-activity/live-metrics/Dags.tsx (hunk 8 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/cluster-activity/live-metrics/Dags.tsx]: Removed thanks

File: airflow/www/static/js/cluster-activity/live-metrics/Dags.tsx (hunk 9 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/cluster-activity/live-metrics/Dags.tsx]: Done

File: airflow/www/static/js/cluster-activity/live-metrics/DagRuns.tsx (hunk 10 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/cluster-activity/live-metrics/DagRuns.tsx]: Done

**Comment by pierrejeambrun** [comment]
@jedcunningham This is how it looks like now, thanks
image

File: airflow/www/static/js/cluster-activity/live-metrics/DagRuns.tsx (hunk 11 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/cluster-activity/live-metrics/DagRuns.tsx]: Under the hood this calls the rest api here `api/v1/dags/~/dagRuns?state=running&limit=5&order_by=start_date`, there is nothing more that the user can't already see.

Indeed it looks like viewers can list all dagruns. Were you thinking of just 'hiding' it in the Cluster Activity page or reworking how the `get_dag_runs` permissions works to not allow this anymore  ?

**Review by eladkal — APPROVED** [review]
I love this PR!

File: airflow/www/static/js/cluster-activity/live-metrics/DagRuns.tsx (hunk 11 above)
Comment (jedcunningham) [reply airflow/www/static/js/cluster-activity/live-metrics/DagRuns.tsx]: My bad, not sure what I was looking at last week. I do see where it's [filtered](https://github.com/apache/airflow/blob/7db42fe6655c28330e80b8a062ef3e07968d6e76/airflow/api_connexion/endpoints/dag_run_endpoint.py#L230) now. I think I might have mistakenly thought this was driven by the changes in `views.py`, which don't have filtering.

So yeah, no concerns here!

**Review by jedcunningham — APPROVED** [review]

**Comment by jedcunningham** [comment]
#protm

**Comment by marclamberti** [comment]
@pierrejeambrun that's amazing 🤩🤩🤩🤩

**Comment by pierrejeambrun** [comment]
Thanks @marclamberti, hope users will like it :)

**Comment by kaxil** [comment]
Great PR, well done @pierrejeambrun 👏

**Comment by vikramkoka** [comment]
This looks amazing @pierrejeambrun, very well done!

**Comment by alex-astronomer** [comment]
Whoaaaaaa...

**Comment by jens-scheffler-bosch** [comment]
@pierrejeambrun I really like the PR and the newly added page!

I just updated my latest master and realize that the menu etry is added in top menu as second entry, so quite present.
As it is some kind of admin related, I think it is better located below the <Browse> or <Admin> menu - it is rather some kind of secondary thing, not a primary entry.

Was there a discussion about how to place it? Would it be okay if I open a PR to move the menu entry?

**Comment by eladkal** [comment]
> Was there a discussion about how to place it? Would it be okay if I open a PR to move the menu entry?

I think the current location is OK.
This is going to be one of the most useful pages. I dont think this page will be soly for users with Admin role.

I suggest to wait after 2.7.0 is released. Gather user feedback and then make changes if needed.
